java-manta-it/src/test/java/com/joyent/manta/client/MantaClientFindIT.java

```diff
@@ -0,0 +1,139 @@
+            List<MantaObject> results = stream.collect(Collectors.toList());
+
+            Assert.assertFalse(results.isEmpty(), "We should have at least one file returned");
+            Assert.assertEquals(results.get(0).getPath(), filePath);
+            Assert.assertFalse(results.get(0).isDirectory());
+        }
+    }
+
+    public void canFindASingleDirectory() throws IOException {
+        String dirPath = testPathPrefix + UUID.randomUUID();
+        mantaClient.putDirectory(dirPath);
+
+        try (Stream<MantaObject> stream = mantaClient.find(testPathPrefix)) {
+            List<MantaObject> results = stream.collect(Collectors.toList());
+
+            Assert.assertFalse(results.isEmpty(), "We should have at least one directory returned");
+            Assert.assertEquals(results.get(0).getPath(), dirPath);
+            Assert.assertTrue(results.get(0).isDirectory());
+        }
+    }
+
+    public void canFindRecursiveDirectoriesAndFiles() throws IOException {
+        List<String> level1Dirs = Arrays.asList(
+                testPathPrefix + UUID.randomUUID(), testPathPrefix + UUID.randomUUID(),
+                testPathPrefix + UUID.randomUUID()
+        );
+
+        List<String> level1Files = Arrays.asList(
+                testPathPrefix + UUID.randomUUID(), testPathPrefix + UUID.randomUUID()
+        );
+
+        for (String dir : level1Dirs) {
+            mantaClient.putDirectory(dir);
+        }
+
+        for (String file : level1Files) {
+            mantaClient.put(file, TEST_DATA, StandardCharsets.UTF_8);
+        }
+
+        List<String> level2Files = level1Dirs
+                .stream()
+                .flatMap(dir -> Stream.of(dir + SEPARATOR + UUID.randomUUID(),
+                                          dir + SEPARATOR + UUID.randomUUID()))
+                .collect(Collectors.toList());
+
+        for (String file : level2Files) {
+            mantaClient.put(file, TEST_DATA, StandardCharsets.UTF_8);
+        }
+
+        int totalObjects = level1Dirs.size() + level1Files.size() + level2Files.size();
+
+        try (Stream<MantaObject> stream = mantaClient.find(testPathPrefix)) {
+            List<MantaObject> results = stream.collect(Collectors.toList());
+
+            Assert.assertFalse(results.isEmpty(),
+                    "We should have many objects returned");
+            Assert.assertEquals(results.size(), totalObjects,
+                    "We should return exactly as many objects as added");
+
+            Stream<MantaObject> resultsStream = results.stream();
+
+            List<String> allObjects = new LinkedList<>();
+            allObjects.addAll(level1Dirs);
+            allObjects.addAll(level1Files);
+            allObjects.addAll(level2Files);
+
+            for (String dir : level1Dirs) {
+                Assert.assertEquals(
+                        resultsStream.filter(obj -> obj.getPath().equals(dir)).count(),
```

Comment:
This test is failing on this line with 

`java.lang.IllegalStateException: stream has already been operated upon or closed`

Moving the call to `results.stream()` into the `for` would probably fix it.